### PR TITLE
Ipv6 privacy extensions

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
#56306

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by lheckemann** [comment]
@GrahamcOfBorg test ipv6

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 1]
```diff
@@ -112,6 +112,19 @@
       </link>
     </para>
    </listitem>
+   <listitem>
+     <para>
+     IPv6 Privacy Extensions are now enabled by default for undeclared
+     interfaces. The previous behaviour was quite misleading — even though
+     the default value for
+     <option>networking.interfaces.*.preferTempAddress</option> was
```
Comment (worldofpeace) [inline nixos/doc/manual/release-notes/rl-1909.xml]:
```suggestion
     <option>networking.interfaces.<name?>.preferTempAddress</option> was
```

Comment (lheckemann) [reply nixos/doc/manual/release-notes/rl-1909.xml]: That's not valid XML AFAIK. I'd prefer leaving it as * to improve readability of the source.

Comment (worldofpeace) [reply nixos/doc/manual/release-notes/rl-1909.xml]: I was trying to make it appear the same as in the option search https://nixos.org/nixos/options.html#networking.interfaces.+prefertempaddress

**Review by fpletz — APPROVED** [review]

**Review by rnhmjoj — CHANGES_REQUESTED** [review]
This works fine but settings `preferTempAddress = false` seems to not work anymore with this change.

**Comment by rnhmjoj** [comment]
I figured out how to make both works by adding a second udev rule:
```patch
diff --git a/nixos/modules/tasks/network-interfaces.nix b/nixos/modules/tasks/network-interfaces.nix
index e14cafa9b12..1a13e386947 100644
--- a/nixos/modules/tasks/network-interfaces.nix
+++ b/nixos/modules/tasks/network-interfaces.nix
@@ -1089,12 +1089,22 @@ in
 
     services.udev.packages =  [
       (pkgs.writeTextFile rec {
-        name = "99-ipv6-privacy-extensions.rules";
-        destination = "/etc/udev/rules.d/99-${name}";
+        name = "ipv6-privacy-extensions.rules";
+        destination = "/etc/udev/rules.d/98-${name}";
         text = ''
+          # enable and prefer IPv6 privacy addresses by default
           ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl net.ipv6.conf.%k.use_tempaddr=2"
         '';
-      }) ] ++ lib.optional (cfg.wlanInterfaces != {})
+      })
+      (pkgs.writeTextFile rec {
+        name = "ipv6-privacy-extensions.rules";
+        destination = "/etc/udev/rules.d/99-${name}";
+        text = concatMapStrings (i: ''
+          # enable IPv6 privacy addresses but prefer EUI-64 addresses for ${i.name}
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl net.ipv6.conf.${i.name}.use_tempaddr=1"
+        '') (filter (i: !i.preferTempAddress) interfaces);
+      })
+    ] ++ lib.optional (cfg.wlanInterfaces != {})
       (pkgs.writeTextFile {
         name = "99-zzz-40-wlanInterfaces.rules";
         destination = "/etc/udev/rules.d/99-zzz-40-wlanInterfaces.rules";
diff --git a/nixos/tests/networking.nix b/nixos/tests/networking.nix
index ed9f287d558..949d946bdc4 100644
--- a/nixos/tests/networking.nix
+++ b/nixos/tests/networking.nix
@@ -510,7 +510,7 @@ let
           '';
         };
       };
-      nodes.client = { pkgs, ... }: with pkgs.lib; {
+      nodes.clientWithPrivacy = { pkgs, ... }: with pkgs.lib; {
         virtualisation.vlans = [ 1 ];
         networking = {
           useNetworkd = networkd;
@@ -522,21 +522,39 @@ let
           };
         };
       };
+      nodes.client = { pkgs, ... }: with pkgs.lib; {
+        virtualisation.vlans = [ 1 ];
+        networking = {
+          useNetworkd = networkd;
+          useDHCP = true;
+          interfaces.eth1 = {
+            preferTempAddress = false;
+            ipv4.addresses = mkOverride 0 [ ];
+            ipv6.addresses = mkOverride 0 [ ];
+          };
+        };
+      };
       testScript = { ... }:
         ''
           startAll;
 
           $client->waitForUnit("network.target");
+          $clientWithPrivacy->waitForUnit("network.target");
           $router->waitForUnit("network-online.target");
 
           # Wait until we have an ip address
+          $clientWithPrivacy->waitUntilSucceeds("ip addr show dev eth1 | grep -q 'fd00:1234:5678:1:'");
           $client->waitUntilSucceeds("ip addr show dev eth1 | grep -q 'fd00:1234:5678:1:'");
 
           # Test vlan 1
+          $clientWithPrivacy->waitUntilSucceeds("ping -c 1 fd00:1234:5678:1::1");
           $client->waitUntilSucceeds("ping -c 1 fd00:1234:5678:1::1");
 
           # Test address used is temporary
-          $client->waitUntilSucceeds("! ip route get fd00:1234:5678:1::1 | grep -q ':[a-f0-9]*ff:fe[a-f0-9]*:'");
+          $clientWithPrivacy->waitUntilSucceeds("! ip route get fd00:1234:5678:1::1 | grep -q ':[a-f0-9]*ff:fe[a-f0-9]*:'");
+
+          # Test address used is EUI-64
+          $client->waitUntilSucceeds("ip route get fd00:1234:5678:1::1 | grep -q ':[a-f0-9]*ff:fe[a-f0-9]*:'");
         '';
     };
     routes = {
```
This patch also updates the test to check both cases.

**Comment by lheckemann** [comment]
Oh, good catch! Thanks.

**Comment by rnhmjoj** [comment]
Soo, can we merge this?

File: nixos/modules/tasks/network-interfaces.nix [hunk 2]
```diff
@@ -1087,7 +1087,24 @@ in
...
     virtualisation.vswitch = mkIf (cfg.vswitches != { }) { enable = true; };
 
-    services.udev.packages = mkIf (cfg.wlanInterfaces != {}) [
+    services.udev.packages =  [
+      (pkgs.writeTextFile rec {
+        name = "ipv6-privacy-extensions.rules";
+        destination = "/etc/udev/rules.d/98-${name}";
+        text = ''
+          # enable and prefer IPv6 privacy addresses by default
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl net.ipv6.conf.%k.use_tempaddr=2"
```
Comment (volth) [inline nixos/modules/tasks/network-interfaces.nix]: It fails if interface name contains `.` (which should be replaced with `/`)

File: nixos/modules/tasks/network-interfaces.nix [hunk 3]
```diff
@@ -1087,7 +1087,24 @@ in
...
+        '';
+      })
+      (pkgs.writeTextFile rec {
+        name = "ipv6-privacy-extensions.rules";
+        destination = "/etc/udev/rules.d/99-${name}";
+        text = concatMapStrings (i: ''
+          # enable IPv6 privacy addresses but prefer EUI-64 addresses for ${i.name}
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl net.ipv6.conf.${i.name}.use_tempaddr=1"
```
Comment (volth) [inline nixos/modules/tasks/network-interfaces.nix]: Here too.
```diff
- ${i.name}
+ ${lib.replaceChars "." "/" i.name}
```

But I do not know how to fix `%k` above

File: nixos/modules/tasks/network-interfaces.nix (hunk 2 above)
Comment (volth) [reply nixos/modules/tasks/network-interfaces.nix]: Sep 06 20:48:27 don1 systemd-udevd[4259]: Process '/nix/store/iysymmqj6c21fdqam8p9lljlly90fxsz-procps-3.3.15/bin/sysctl net.ipv6.conf.tinc.dmz.use_tempaddr=2' failed with exit code 255.

File: nixos/modules/tasks/network-interfaces.nix (hunk 3 above)
Comment (lheckemann) [reply nixos/modules/tasks/network-interfaces.nix]: Maybe by writing to `/proc/sys/net/ipv6/conf/%k/use_tempaddr` instead of using sysctl…

**Pull request merged**